text/0162-refactoring-support.md

```diff
@@ -0,0 +1,702 @@
+didn't detect them automatically. This may happen a resource is moved and
+modified at the same time, for example. You may also be in a situation in which
+the CLI detected a refactor, but you actually intend to replace it. To
+accomplish this, you can remove the unwanted mapping from the mapping file.
+
+### Settings
+
+You have a few settings available to control the behavior of the refactoring
+feature.
+
+For both `deploy` and `refactor`:
+
+- `--record-resource-mapping=<FILE>`: writes the mapping to a file. The file can
+  be used later to apply the same refactors in other environments.
+- `--resource-mapping=<FILE>`: use the mapping from a file, instead of computing
+  it. This file can be generated using the `--record-resource-mapping` option.
+- `--unstable=refactor`: used to acknowledge that this feature is experimental.
+  If the flag is not set, and the CLI would try to perform some refactor, the
+  command fails with an error message explaining why.
+
+For `deploy` only:
+
+- `--refactoring-action=[ACTION]`: the action to take in case there is a
+  refactor to be made. Possible values for `ACTION` are:
+    - `refactor`: automatically refactor and deploy.
+    - `skip`: deploy without refactoring. This is the default value.
+
+For `refactor` only:
+
+- `--dry-run`: prints the mapping to the console, but does not apply it.
+
+All these settings are also available in the `cdk.json` file:
+
+```json
+{
+  "app": "...",
+  "refactor": {
+    "refactoringAction": "refactor",
+    "exportMapping": "output.json",
+    "dryRun": true
+  }
+}
+```
+
+### Rollback
+
+After refactoring the stack, the CLI will proceed with the deployment
+(assuming that is your choice). If the deployment fails, and CloudFormation
+rolls it back, the CLI will execute a second refactor, in reverse, to bring the
+resources back to their original locations.
+
+If you don't want the CLI to perform the rollback refactor, you can use the
+`--no-rollback` flag, which also controls the rollback behavior of the
+deployment.
+
+### Programmatic access
+
+The same refactoring feature is also available in the CDK toolkit library:
+
+```typescript
+declare const toolkit: Toolkit;
+declare const cx: ICloudAssemblySource;
+
+// To execute possible refactors as part of the deploy operation:
+await toolkit.deploy(cx, {
+  refactoring: RefactoringMode.EXECUTE_AND_DEPLOY
+});
+
+// Or, if you just want to refactor the stacks:
+await toolkit.refactor(cx);
+```
+
+### Limitations and failure modes
+
+#### Pipelines with version superseding
+
+As we have seen, there are two ways to apply refactors: automatically, by
+consuming a mapping file, or interactively, when the CLI computes the refactor
+to be made, and the user is asked whether to proceed. If you want to use this
+feature in a CI/CD pipeline, so that the refactors are applied automatically,
+you must use commit the mapping file along with your code and use it on each
```

Comment:
Will the mapping file be used in the next refactoring, or is it generated without considering the history? What happens if a customer refactors the same resource twice?